oss_client.cpp

```diff
@@ -0,0 +1,907 @@
+                  m_unix_sockaddr ? sizeof(struct sockaddr_un) : addr.ci_addrlen) == -1) {
+        if (errno == EINPROGRESS || errno == EWOULDBLOCK)
+            return 0;
+        benchmark_error_log("connect failed, error = %s\n", strerror(errno));
+        return -1;
+    }
+
+    return 0;
+}
+
+int oss_client::add_connection(int min_slots_range, int max_slots_range, char* addr, char* port) {
+    shard_connection* conn = new shard_connection();
+    conn->setup_connection(m_connections.size(), this, MAIN_CONNECTION->protocol);
+    m_connections.push_back(conn);
+
+    // set slots range
+    conn->min_slots_range = min_slots_range;
+    conn->max_slots_range = max_slots_range;
+
+    // set socket
+    struct connect_info ci;
+    struct addrinfo *addr_info;
+    struct addrinfo hints;
+    int ret;
+
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_flags = AI_PASSIVE;
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_family = AF_INET;
+
+    int res = getaddrinfo(addr, port, &hints, &addr_info);
+    if (res != 0) {
+        benchmark_error_log("connect: resolve error: %s\n", gai_strerror(res));
+        return -1;
+    }
+
+    ci.ci_family = addr_info->ai_family;
+    ci.ci_socktype = addr_info->ai_socktype;
+    ci.ci_protocol = addr_info->ai_protocol;
+    assert(addr_info->ai_addrlen <= sizeof(ci.addr_buf));
+    memcpy(ci.addr_buf, addr_info->ai_addr, addr_info->ai_addrlen);
+    ci.ci_addr = (struct sockaddr *) ci.addr_buf;
+    ci.ci_addrlen = addr_info->ai_addrlen;
+
+    ret = conn->setup_socket(&ci);
+    if (ret < 0) {
+        benchmark_error_log("connect: failed to setup socket\n");
+        return -1;
+    }
+
+    // set up read/write events
+    conn->setup_events(m_event_base);
+
+    // call connect
+    if (::connect(conn->sockfd,
+                  m_unix_sockaddr ? (struct sockaddr *) m_unix_sockaddr : ci.ci_addr,
+                  m_unix_sockaddr ? sizeof(struct sockaddr_un) : ci.ci_addrlen) == -1) {
+        if (errno == EINPROGRESS || errno == EWOULDBLOCK)
+            return 0;
+        benchmark_error_log("connect failed, error = %s\n", strerror(errno));
+        return -1;
+    }
+
+    // check connection
+    int error = -1;
+    socklen_t errsz = sizeof(error);
+
+    if (getsockopt(conn->sockfd, SOL_SOCKET, SO_ERROR, (void *) &error, &errsz) == -1) {
+        benchmark_error_log("connect: error getting connect response (getsockopt): %s\n", strerror(errno));
+        return -1;
+    }
+
+    if (error != 0) {
+        benchmark_error_log("connect: connection failed: %s\n", strerror(error));
+        return -1;
+    }
+
+    return 0;
+}
+
+bool oss_client::is_conn_setup_done(void)
+{
+    if (client::is_conn_setup_done())
+        return m_cluster_slots_state == slots_done;
+    else
+        return false;
+}
+
+bool oss_client::send_conn_setup_commands(struct timeval timestamp)
+{
+    bool sent = false;
+
+    if (m_config->authenticate && m_authentication != auth_done) {
```

Comment:
Unless I'm missing something here, the authentication and DB selection are always done on the main connection and never on the other connections.

In fact, the authentication and selected DB state belong to the shard connection rather than the client.